src/compute-client/src/controller/instance.rs

```diff
@@ -1570,12 +1550,20 @@ where
         }
     }
 
+    /// Handle new frontiers, returning any compute response that needs to
+    /// be sent to the client.
     fn handle_frontier_upper(
         &mut self,
         id: GlobalId,
         new_frontier: Antichain<T>,
         replica_id: ReplicaId,
-    ) {
+    ) -> Option<ComputeControllerResponse<T>> {
+        let old_upper = self
+            .compute
+            .collection(id)
+            .ok()
+            .map(|state| state.write_frontier.clone());
```

Comment:
We are getting the collection in the next `let-else`, so if we move this block down we don't have to handle the absence case.